### PR TITLE
NS directive

### PR DESCRIPTION
**Pull request by dpisarenko**
Issue: https://github.com/yegor256/xembly/issues/4

### CODE REVIEW

**Comment by davvd** [comment]
Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@krzyk review this one,please

**Comment by krzyk** [comment]
@dpisarenko travis build doesn't pass, there are checkstyle errors

File: src/test/java/org/xembly/NsDirectiveTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * @version $Id$
+ * @since 1.0
+ */
+public final class NsDirectiveTest {
+    /**
+     * Verifies that NsDirective appends a namespace to a node.
+     * @throws Exception Thrown in case of an error.
+     */
+    @Test
+    public void exec() throws Exception {
```
Comment (krzyk) [inline src/test/java/org/xembly/NsDirectiveTest.java]: please give a more descriptive method name (see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names)

File: src/main/java/org/xembly/NsDirective.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    @Override
+    public Pointer exec(final Node dom, final Pointer ptr, final Stack stack) {
+        try {
+            final AttrDirective attr = new AttrDirective(
+                "xmlns",
+                this.namespace
+            );
+            return attr.exec(dom, ptr, stack);
+        } catch (final XmlContentException exception) {
+            exception.printStackTrace();
```
Comment (krzyk) [inline src/main/java/org/xembly/NsDirective.java]: please remove printing stack trace, just throw `IllegalArgumentException` with the source exception

File: src/main/java/org/xembly/NsDirective.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import com.jcabi.aspects.Immutable;
+import lombok.EqualsAndHashCode;
+import org.w3c.dom.Node;
+
+/**
+ * Namespace directive.
+ * @author Dmitri Pisarenko (dp@altruix.co)
+ * @version $Id$
+ * @since 1.0
```
Comment (krzyk) [inline src/main/java/org/xembly/NsDirective.java]: this should be `@since 0.19.3`

File: src/test/java/org/xembly/NsDirectiveTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.junit.Test;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+
+/**
+ * Validates NsDirective class.
+ * @author Dmitri Pisarenko (dp@altruix.co)
+ * @version $Id$
+ * @since 1.0
```
Comment (krzyk) [inline src/test/java/org/xembly/NsDirectiveTest.java]: here also `@since 0.19.3`

**Comment by krzyk** [comment]
@dpisarenko  and few comments above

**Comment by dpisarenko** [comment]
@krzyk I fixed the issues you mentioned, thanks. Please review the modified code.

**Comment by krzyk** [comment]
@rultor merge pls

File: src/main/java/org/xembly/NsDirective.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+     * @param nsp Namespace, which should be added to a node.
+     */
+    public NsDirective(final String nsp) {
+        this.namespace = nsp;
+    }
+
+    @Override
+    public String toString() {
+        return String.format(
+            "NS %s",
```
Comment (yegor256) [inline src/main/java/org/xembly/NsDirective.java]: I think it's wrong. Imagine we have this namespace: `"abc; 'cde'"`

**Comment by yegor256** [comment]
@dpisarenko look at other directives, they are using `Arg` instead of `String`

File: src/main/java/org/xembly/NsDirective.java (hunk 5 above)
Comment (dpisarenko) [inline src/main/java/org/xembly/NsDirective.java]: @yegor256 What method is wrong - `toString` or the constructor?

Comment (yegor256) [inline src/main/java/org/xembly/NsDirective.java]: the ctor. you encapsulate `String` and it's strong. see how `AttrDirective` is done, for example

File: src/test/java/org/xembly/NsDirectiveTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            .newDocumentBuilder().newDocument();
+        final Element root = dom.createElement("f");
+        dom.appendChild(root);
+        new NsDirective("somens").exec(
+            dom, new DomPointer(Collections.<Node>singletonList(root)),
+            new DomStack()
+        );
+        MatcherAssert.assertThat(
+            new XMLDocument(dom).toString(),
+            org.hamcrest.Matchers.containsString("<f xmlns=\"somens\"/>")
```
Comment (yegor256) [inline src/test/java/org/xembly/NsDirectiveTest.java]: why full class name here?

File: src/test/java/org/xembly/NsDirectiveTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    public void setsNsAttr() throws Exception {
+        final Document dom = DocumentBuilderFactory.newInstance()
+            .newDocumentBuilder().newDocument();
+        final Element root = dom.createElement("f");
+        dom.appendChild(root);
+        new NsDirective("somens").exec(
+            dom, new DomPointer(Collections.<Node>singletonList(root)),
+            new DomStack()
+        );
+        MatcherAssert.assertThat(
```
Comment (yegor256) [inline src/test/java/org/xembly/NsDirectiveTest.java]: try this test instead:

```
MatcherAssert.assertThat(
  new XMLDocument(dom),
  XhtmlMatchers.hasXPath("/ns1:f", "somens")
);
```

**Comment by dpisarenko** [comment]
@yegor256 Thanks. I fixed the errors. Please review again.

**Comment by yegor256** [comment]
@dpisarenko I think we should also update `index.apt.vm`. First, we should mention this new `NS` directive there. Second, we should remove the "not implemented yet" line at the bottom.

**Comment by dpisarenko** [comment]
@yegor256 I updated `index.apt.vm`. Please review again. Thanks.

**Comment by yegor256** [comment]
@rultor merge this pls

**Comment by davvd** [comment]
@krzyk I added **17 mins** to your account, many thanks for working with the project!. +17 added to your rating, current score is: [+2896](http://www.netbout.com/b/35035?open=rating)

**Comment by davvd** [comment]
@rultor deploy

**Pull request merged**